File: sdks/official/fraiseql-java/src/test/java/com/fraiseql/core/GenerateParitySchema.java
package com.fraiseql.core;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import org.junit.jupiter.api.Test;

import java.io.File;

/**
 * Generate parity schema for cross-SDK comparison.
 *
 * Produces the canonical parity-schema.json in array format compatible with
 * Python / TypeScript / Go / PHP output and the compare_parity_schemas.py script.
 *
 * Run via Maven:
 *   mvn -q test -Dtest=GenerateParitySchema -DschemaOutputFile=/tmp/parity-java.json
 *
 * If schemaOutputFile is not set, the JSON is printed to stdout.
 */
public class GenerateParitySchema {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    @Test
    void generateAndExport() throws Exception {
        ObjectNode root = MAPPER.createObjectNode();

        // ── Types ──────────────────────────────────────────────────────────
        ArrayNode types = MAPPER.createArrayNode();

        types.add(makeType("User", "v_user", false,
            makeField("id",    "ID",     false),
            makeField("email", "String", false),
            makeField("name",  "String", false)
        ));

        types.add(makeType("Order", "v_order", false,
            makeField("id",    "ID",    false),
            makeField("total", "Float", false)
        ));

        ObjectNode userNotFound = makeType("UserNotFound", "v_user_not_found", false,
            makeField("message", "String", false),
            makeField("code",    "String", false)
        );
        userNotFound.put("is_error", true);
        types.add(userNotFound);

        root.set("types", types);

        // ── Queries ────────────────────────────────────────────────────────
        ArrayNode queries = MAPPER.createArrayNode();

        ObjectNode users = MAPPER.createObjectNode();
        users.put("name", "users");
        users.put("return_type", "User");
        users.put("returns_list", true);
        users.put("nullable", false);
        users.put("sql_source", "v_user");
        users.set("arguments", MAPPER.createArrayNode());
        queries.add(users);

        ObjectNode tenantOrders = MAPPER.createObjectNode();
        tenantOrders.put("name", "tenantOrders");
        tenantOrders.put("return_type", "Order");
        tenantOrders.put("returns_list", true);
        tenantOrders.put("nullable", false);
        tenantOrders.put("sql_source", "v_order");
        tenantOrders.set("inject_params", makeInjectParam("tenant_id", "jwt", "tenant_id"));
        tenantOrders.put("cache_ttl_seconds", 300);
        tenantOrders.put("requires_role", "admin");
        tenantOrders.set("arguments", MAPPER.createArrayNode());
        queries.add(tenantOrders);

        root.set("queries", queries);

        // ── Mutations ──────────────────────────────────────────────────────
        ArrayNode mutations = MAPPER.createArrayNode();

        ObjectNode createUser = MAPPER.createObjectNode();
        createUser.put("name", "createUser");
        createUser.put("return_type", "User");
        createUser.put("sql_source", "fn_create_user");
        createUser.put("operation", "insert");
        ArrayNode createUserArgs = MAPPER.createArrayNode();
        createUserArgs.add(makeArgument("email", "String", false));
        createUserArgs.add(makeArgument("name",  "String", false));
        createUser.set("arguments", createUserArgs);
        mutations.add(createUser);

        ObjectNode placeOrder = MAPPER.createObjectNode();
        placeOrder.put("name", "placeOrder");
        placeOrder.put("return_type", "Order");
        placeOrder.put("sql_source", "fn_place_order");
        placeOrder.put("operation", "insert");
        placeOrder.set("inject_params", makeInjectParam("user_id", "jwt", "sub"));
        ArrayNode invalidViews = MAPPER.createArrayNode();
        invalidViews.add("v_order_summary");
        placeOrder.set("invalidates_views", invalidViews);
        ArrayNode invalidTables = MAPPER.createArrayNode();
        invalidTables.add("tf_sales");
        placeOrder.set("invalidates_fact_tables", invalidTables);
        placeOrder.set("arguments", MAPPER.createArrayNode());
        mutations.add(placeOrder);

        root.set("mutations", mutations);

        // ── Output ─────────────────────────────────────────────────────────
        String json = MAPPER.writerWithDefaultPrettyPrinter().writeValueAsString(root);
        String outputFile = System.getProperty("schemaOutputFile");
        if (outputFile != null && !outputFile.isEmpty()) {
            MAPPER.writerWithDefaultPrettyPrinter().writeValue(new File(outputFile), root);
        } else {
            System.out.println(json);
        }
    }

    // ── Helpers ──────────────────────────────────────────────────────────────

    private static ObjectNode makeType(String name, String sqlSource, boolean isError,
                                       ObjectNode... fields) {
        ObjectNode t = MAPPER.createObjectNode();
        t.put("name", name);
        t.put("sql_source", sqlSource);
        if (isError) {
            t.put("is_error", true);
        }
        ArrayNode fa = MAPPER.createArrayNode();
        for (ObjectNode f : fields) {
            fa.add(f);
        }
        t.set("fields", fa);
        return t;
    }

    private static ObjectNode makeField(String name, String type, boolean nullable) {
        ObjectNode f = MAPPER.createObjectNode();
        f.put("name", name);
        f.put("type", type);
        f.put("nullable", nullable);
        return f;
    }

    private static ObjectNode makeArgument(String name, String type, boolean nullable) {
        ObjectNode a = MAPPER.createObjectNode();
        a.put("name", name);
        a.put("type", type);
        a.put("nullable", nullable);
        return a;
    }

    /** Build: {"<param>": "<source>:<claim>"} — matches Python SDK format. */
    private static ObjectNode makeInjectParam(String param, String source, String claim) {
        ObjectNode ip = MAPPER.createObjectNode();
        ip.put(param, source + ":" + claim);
        return ip;
    }
}

File: sdks/official/fraiseql-ruby/test/generate_parity_schema.rb
# frozen_string_literal: true

# Generate parity schema for cross-SDK comparison.
#
# Produces the canonical parity-schema JSON compatible with the Python
# reference generator and compare_schemas.py.
#
# Usage:
#   ruby test/generate_parity_schema.rb
#
# When SCHEMA_OUTPUT_FILE is set the JSON is written to that path instead
# of stdout:
#   SCHEMA_OUTPUT_FILE=/tmp/schema_ruby.json ruby test/generate_parity_schema.rb

require "json"

# ── Types ──────────────────────────────────────────────────────────────────

def make_field(name, type, nullable)
  { "name" => name, "type" => type, "nullable" => nullable }
end

def make_argument(name, type, nullable)
  { "name" => name, "type" => type, "nullable" => nullable }
end

types = [
  {
    "name" => "User",
    "sql_source" => "v_user",
    "fields" => [
      make_field("id",    "ID",     false),
      make_field("email", "String", false),
      make_field("name",  "String", false),
    ],
  },
  {
    "name" => "Order",
    "sql_source" => "v_order",
    "fields" => [
      make_field("id",    "ID",    false),
      make_field("total", "Float", false),
    ],
  },
  {
    "name" => "UserNotFound",
    "sql_source" => "v_user_not_found",
    "is_error" => true,
    "fields" => [
      make_field("message", "String", false),
      make_field("code",    "String", false),
    ],
  },
]

# ── Queries ────────────────────────────────────────────────────────────────

queries = [
  {
    "name"         => "users",
    "return_type"  => "User",
    "returns_list" => true,
    "nullable"     => false,
    "sql_source"   => "v_user",
    "arguments"    => [],
  },
  {
    "name"              => "tenantOrders",
    "return_type"       => "Order",
    "returns_list"      => true,
    "nullable"          => false,
    "sql_source"        => "v_order",
    "inject_params"     => { "tenant_id" => "jwt:tenant_id" },
    "cache_ttl_seconds" => 300,
    "requires_role"     => "admin",
    "arguments"         => [],
  },
]

# ── Mutations ──────────────────────────────────────────────────────────────

mutations = [
  {
    "name"        => "createUser",
    "return_type" => "User",
    "sql_source"  => "fn_create_user",
    "operation"   => "insert",
    "arguments"   => [
      make_argument("email", "String", false),
      make_argument("name",  "String", false),
    ],
  },
  {
    "name"                    => "placeOrder",
    "return_type"             => "Order",
    "sql_source"              => "fn_place_order",
    "operation"               => "insert",
    "inject_params"           => { "user_id" => "jwt:sub" },
    "invalidates_views"       => ["v_order_summary"],
    "invalidates_fact_tables" => ["tf_sales"],
    "arguments"               => [],
  },
]

# ── Output ─────────────────────────────────────────────────────────────────

schema = { "types" => types, "queries" => queries, "mutations" => mutations }
json = JSON.pretty_generate(schema)

output_file = ENV["SCHEMA_OUTPUT_FILE"]
if output_file && !output_file.empty?
  File.write(output_file, json)
else
  puts json
end
